### PR TITLE
XCM fee payment and weight limits

### PR DESCRIPTION
**Pull request by gavofyork**
- Make the `Xcm` and `Order` enums generic over the `Call` type.
- Introduce `Order::BuyExecution` 
- Introduce additional traits `WeightTrader`, `ShouldExecute` and `WeightBounds`.
- Introduce basic implementions of each trait, including several sensible ones for `ShouldExecute`.
- Introduce a mock test  environment and plenty of tests.

TODO: 
- [ ] Rename Xcm -> opaque::Xcm and rename XcmGeneric -> generic::Xcm, same for `Order` and `VersionedXcm`
- [ ] Move builder structs in xcm-executor over into xcm-builder
- [ ] More tests in xcm-executor/xcm-builder for new logic

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1754,7 +1658,7 @@ checksum = "3f9eec918d3f24069decb9af1554cad7c880e2da24a9afd88aca000531ab82c1"
 [[package]]
 name = "fork-tree"
 version = "3.0.0"
-source = "git+https://github.com/paritytech/substrate#b5978187b9c9fbfcf535ab96161a92c0b330540b"
+source = "git+https://github.com/paritytech/substrate?branch=rococo-v1#401c24e8a62cdf058882b0e92815faef966d9fa1"
```
Comment (ordian) [inline Cargo.lock:1661]: that's probably not intended

**Review by shawntabrizi — APPROVED** [review]

**Review by rphmeier — APPROVED** [review]

File: xcm/src/v0/mod.rs [hunk 2]
```diff
@@ -131,33 +131,11 @@ pub enum Xcm {
...
-	/// - `inner: VersionedXcm`: The message to be wrapped and relayed.
-	///
-	/// Safety: No concerns.
-	///
-	/// Kind: *Instruction*.
-	///
-	/// Errors:
-	RelayTo { dest: MultiLocation, inner: Box<VersionedXcm> },
+	/// Unused
+	Unused5,
```
Comment (rphmeier) [inline xcm/src/v0/mod.rs:135]: beyond the scope, of the PR but I'd expect explicit `#[codec(index = X)]` attrs

**Comment by rphmeier** [comment]
Looks like this PR has grown beyond removing stale XCM messages

**Comment by xlc** [comment]
Does this means all the XCM executed on relaychain needs to pay relaychain token as fee? Also the parachain needs to provide a reasonable estimate upfront to make sure it buy enough weights for the execution?

**Comment by gavofyork** [comment]
> Does this means all the XCM executed on relaychain needs to pay relaychain token as fee?

It probably will, yes. For system and common-good parachains which are trusted by the relay-chain, then it's not unreasonable to expect them to get relay-chain XCM computation for free.

> Also the parachain needs to provide a reasonable estimate upfront to make sure it buy enough weights for the execution?

Yes. If the weight fees end up being too volatile to be practical, then we'll have to look into introducing a damping mechnanism.

**Pull request closed without merging**